pandas/period.pyx

```diff
@@ -0,0 +1,1130 @@
+    """
+    if mult == 1:
+        return period_ord_w_mult
+
+    return period_ord_w_mult * mult + 1;
+
+def dt64arr_to_periodarr(ndarray[int64_t] dtarr, int freq, tz=None):
+    """
+    Convert array of datetime64 values (passed in as 'i8' dtype) to a set of
+    periods corresponding to desired frequency, per period convention.
+    """
+    cdef:
+        ndarray[int64_t] out
+        Py_ssize_t i, l
+        pandas_datetimestruct dts
+
+    l = len(dtarr)
+
+    out = np.empty(l, dtype='i8')
+
+    if tz is None:
+        for i in range(l):
+            if dtarr[i] == iNaT:
+                out[i] = iNaT
+                continue
+            pandas_datetime_to_datetimestruct(dtarr[i], PANDAS_FR_ns, &dts)
+            out[i] = get_period_ordinal(dts.year, dts.month, dts.day,
+                                        dts.hour, dts.min, dts.sec, dts.us, dts.ps, freq)
+    else:
+        out = localize_dt64arr_to_period(dtarr, freq, tz)
+    return out
+
+def periodarr_to_dt64arr(ndarray[int64_t] periodarr, int freq):
+    """
+    Convert array to datetime64 values from a set of ordinals corresponding to
+    periods per period convention.
+    """
+    cdef:
+        ndarray[int64_t] out
+        Py_ssize_t i, l
+
+    l = len(periodarr)
+
+    out = np.empty(l, dtype='i8')
+
+    for i in range(l):
+        if periodarr[i] == iNaT:
+            out[i] = iNaT
+            continue
+        out[i] = period_ordinal_to_dt64(periodarr[i], freq)
+
+    return out
+
+cdef char START = 'S'
+cdef char END = 'E'
+
+cpdef int64_t period_asfreq(int64_t period_ordinal, int freq1, int freq2,
+                            bint end):
+    """
+    Convert period ordinal from one frequency to another, and if upsampling,
+    choose to use start ('S') or end ('E') of period.
+    """
+    cdef:
+        int64_t retval
+
+    if period_ordinal == iNaT:
+        return iNaT
+
+    if end:
+        retval = asfreq(period_ordinal, freq1, freq2, END)
+    else:
+        retval = asfreq(period_ordinal, freq1, freq2, START)
+
+    if retval == INT32_MIN:
+        raise ValueError('Frequency conversion failed')
+
+    return retval
+
+def period_asfreq_arr(ndarray[int64_t] arr, int freq1, int freq2, bint end):
+    """
+    Convert int64-array of period ordinals from one frequency to another, and
+    if upsampling, choose to use start ('S') or end ('E') of period.
+    """
+    cdef:
+        ndarray[int64_t] result
+        Py_ssize_t i, n
+        freq_conv_func func
+        asfreq_info finfo
+        int64_t val, ordinal
+        char relation
+
+    n = len(arr)
+    result = np.empty(n, dtype=np.int64)
+
+    func = get_asfreq_func(freq1, freq2)
+    get_asfreq_info(freq1, freq2, &finfo)
+
+    if end:
+        relation = END
+    else:
+        relation = START
+
+    mask = arr == iNaT
+    if mask.any():      # NaT process
+        for i in range(n):
+            val = arr[i]
+            if val != iNaT:
+                val = func(val, relation, &finfo)
```

Comment:
just combine these into a single loop
In python this matters but not in cython where these are fast comparisons (you can do a mini profile to see if it matters in any event) - it's just more clear in the code